node/src/components/chainspec_loader.rs

```diff
@@ -231,42 +216,371 @@ impl ChainspecLoader {
+    pub(crate) fn genesis_state_root_hash(&self) -> Option<Digest> {
+        if self.chainspec.is_genesis() {
+            return self.state_root_hash;
+        }
+        None
     }
 
-    pub(crate) fn chainspec(&self) -> &Chainspec {
+    pub(crate) fn chainspec(&self) -> &Arc<Chainspec> {
         &self.chainspec
     }
+
+    pub(crate) fn next_upgrade(&self) -> Option<NextUpgrade> {
+        self.next_upgrade.clone()
+    }
+
+    pub(crate) fn highest_block_hash(&self) -> Option<BlockHash> {
+        self.highest_block_hash
+    }
+
+    fn handle_initialize<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        maybe_cached_protocol_version: Option<Version>,
+        maybe_block: Option<Box<Block>>,
+    ) -> Effects<Event>
+    where
+        REv: From<Event> + From<StateStoreRequest> + From<ContractRuntimeRequest> + Send,
+    {
+        // Cache the current protocol version.
+        let mut effects = effect_builder
+            .save_state(
+                STORAGE_KEY.as_bytes().into(),
+                self.chainspec.protocol_config.version.clone(),
+            )
+            .ignore();
+
+        let highest_block = match maybe_block {
+            Some(block) => {
+                self.highest_block_hash = Some(*block.hash());
+                block
+            }
+            None => {
+                // This is an initial run since we have no blocks.
+                if self.chainspec.is_genesis() {
+                    // This is a valid initial run on a new network at genesis.
+                    trace!("valid initial run at genesis");
+                    effects.extend(
+                        effect_builder
+                            .commit_genesis(Arc::clone(&self.chainspec))
+                            .event(Event::CommitGenesisResult),
+                    );
+                    return effects;
+                } else {
+                    // This is an invalid run of a node version issued after genesis.  Instruct the
+                    // process to exit and downgrade the version.
+                    warn!(
+                        "invalid run, no blocks stored but not a genesis chainspec: exit to \
+                        downgrade"
+                    );
+                    self.reactor_exit =
+                        Some(ReactorExit::ProcessShouldExit(ExitCode::DowngradeVersion));
+                    return Effects::new();
+                }
+            }
+        };
+        let highest_block_era_id = highest_block.header().era_id().0;
+
+        let cached_protocol_version = maybe_cached_protocol_version
+            .expect("cached protocol version must exist if we have a stored block");
+        let current_chainspec_activation_point =
+            self.chainspec.protocol_config.activation_point.era_id.0;
+
+        if highest_block_era_id + 1 == current_chainspec_activation_point {
```

Comment:
You could keep them as `EraId` type:
```suggestion
        if highest_block_era_id.successor() == current_chainspec_activation_point {
```

Reply:
Nice - thanks!